src/Framework/Migrations/Controllers/BatchMigrationRunner.php

```diff
@@ -0,0 +1,214 @@
+<?php
+
+namespace Give\Framework\Migrations\Controllers;
+
+use ActionScheduler_Store;
+use Exception;
+use Give\Framework\Database\DB;
+use Give\Framework\Migrations\Contracts\BatchMigration;
+use Give\MigrationLog\MigrationLogStatus;
+
+/**
+ * Batch Migration runner controller
+ *
+ * @unreleased
+ */
+class BatchMigrationRunner
+{
+    /**
+     * @var BatchMigration
+     */
+    private $migration;
+
+    /**
+     * @var ActionScheduler_Store
+     */
+    private $actionSchedulerStore;
+
+    /**
+     * @unreleased
+     */
+    public function __construct(BatchMigration $migration)
+    {
+        $this->migration = $migration;
+        $this->actionSchedulerStore = ActionScheduler_Store::instance();
+    }
+
+    /**
+     * Run batch migration
+     *
+     * @unreleased
+     *
+     * @throws Exception
+     */
+    public function run(): string
+    {
+        $this->registerAction();
+
+        $actions = $this->actionSchedulerStore->query_actions([
+            'group' => $this->getGroup(),
+            'per_page' => 0,
+        ]);
+
+        if ( ! $actions) {
+            $itemsCount = $this->migration->getItemsCount();
+
+            // Bailout if there are no items to process
+            if ( ! $itemsCount) {
+                return MigrationLogStatus::SUCCESS;
+            }
+
+            $current = 0;
+            $batches = ceil($itemsCount / $this->migration->getBatchSize());
+
+            // Register migration action for each batch
+            for ($i = 0; $i < $batches; $i++) {
+                if ($items = $this->migration->getBatchItemsAfter($current)) {
+                    [$firstId, $lastId] = $items;
```

Comment:
Let's add a validation step here that throws an exception if it's not a 2-element array with integers.

Reply:
The thing with the last ID parameter is that it will be null on the last step (if there is more data). Check https://github.com/impress-org/givewp/blob/24162f82fd016d24b183407c48c0b7d98d79c0fd/src/Framework/Migrations/Controllers/BatchMigrationRunner.php#L158-L159

We are pulling the last ID from the last action. That last ID will be the first ID of the new action. We don't need the last ID for new action because we will process all the items we found. See https://github.com/impress-org/givewp/blob/24162f82fd016d24b183407c48c0b7d98d79c0fd/src/Campaigns/Migrations/Donations/AddCampaignId.php#L77

Reply:
The good news is we know when it should be what based on the iteration of the loop, and can therefore validate it accordingly. So:
* If $i is less than $batches, it should be a 2-index array of integers
* If $i is equal to $batches, it should be null

Reply:
That's true when we set the initial number of batches. What is a bit different is that if the `hasMoreItemsToBatch` function returns true, the BatchMigrationRunner will schedule another batch where all remaining items will be processed. In that situation we don't need to know the range of IDs, we just need to know the last processed item ID. Also, actions/batches don't even know what is the current batch number, they just have the range of IDs for which they are responsible.

Reply:
Fair enough. If we really just don't have a way of validating here, then YOLO. 🚬 🧐

Reply:
What really matters is that we pass the correct IDs, the batch number is not even important. It's used to determine how many batches we should have at that point in time.